publish.go

```diff
@@ -0,0 +1,326 @@
+	log.Debugf("Checking if updating listing is required, whats new dir is '%v'", configs.WhatsnewsDir)
+	if configs.WhatsnewsDir != "" {
+		fmt.Println()
+		log.Infof("Update listing started")
+
+		recentChangesMap, err := readLocalisedRecentChanges(configs.WhatsnewsDir)
+		if err != nil {
+			return fmt.Errorf("failed to read whatsnews, error: %s", err)
+		}
+
+		var releaseNotes []*androidpublisher.LocalizedText
+		for language, recentChanges := range recentChangesMap {
+			releaseNotes = append(releaseNotes, &androidpublisher.LocalizedText{
+				Language:        language,
+				Text:            recentChanges,
+				ForceSendFields: []string{},
+				NullFields:      []string{},
+			})
+		}
+		release.ReleaseNotes = releaseNotes
+		log.Infof("Update listing finished")
+	}
+	return nil
+}
+
+// readLocalisedRecentChanges reads the recent changes from the given path and returns them as a map.
+func readLocalisedRecentChanges(recentChangesDir string) (map[string]string, error) {
+	recentChangesMap := map[string]string{}
+
+	pattern := filepath.Join(recentChangesDir, "whatsnew-*-*")
+	recentChangesPaths, err := filepath.Glob(pattern)
+	if err != nil {
+		return map[string]string{}, err
+	}
+
+	pattern = `whatsnew-(?P<local>.*-.*)`
+	re := regexp.MustCompile(pattern)
+
+	for _, recentChangesPath := range recentChangesPaths {
+		matches := re.FindStringSubmatch(recentChangesPath)
+		if len(matches) == 2 {
+			language := matches[1]
+			content, err := fileutil.ReadStringFromFile(recentChangesPath)
+			if err != nil {
+				return map[string]string{}, err
+			}
+
+			recentChangesMap[language] = content
+		}
+	}
+	if len(recentChangesMap) > 0 {
+		log.Debugf("Found the following recent changes:")
+		for language, recentChanges := range recentChangesMap {
+			log.Debugf("%v\n", language)
+			log.Debugf("Content: %v", recentChanges)
+		}
+	} else {
+		log.Debugf("No recent changes found")
+	}
+
+	return recentChangesMap, nil
+}
+
+// getRelease gets a release from a track based on it's status. Note we would get error if we would like to have
+// multiple instances of a release with the same status. Example: "error: googleapi: Error 400: Too many completed
+// releases specified., releasesTooManyCompletedReleases".
+func getRelease(userFraction float64, releases *[]*androidpublisher.TrackRelease) *androidpublisher.TrackRelease {
+	status := releaseStatusFromConfig(userFraction)
+	for _, release := range *releases {
+		if status == release.Status {
+			return release
+		}
+	}
+	newRelease := androidpublisher.TrackRelease{
+		Status: status,
+	}
+	if userFraction != 0 {
+		newRelease.UserFraction = userFraction
+	}
+	*releases = append(*releases, &newRelease)
+	return &newRelease
+}
+
+// updateRelease creates and returns a new release object with the given version codes.
+func updateRelease(configs Configs, versionCodes googleapi.Int64s, release *androidpublisher.TrackRelease) error {
```

Comment:
Please test case when UntrackBlockingVersions == false, and no regression when true.
Would merge with caller function (updatetracks) for readibility.

Reply:
I would not merge because updating track and release should be two separate functions, as multiple use cases are possible.

Reply:
Currently this function does multiple things, and is hard to understand at a glance what is it doing. The naming also contributes, and also that it adds 1 more layer of function calls.

Reply:
main -> updateTrack -> updateRelease : the extra layer is not a big deal

Agree with you, the naming could be better, as it does 2 things. Updated doc and name.